utilities/simulator_cache/cache_simulator.h

```diff
@@ -38,22 +61,102 @@ class CacheSimulator {
   CacheSimulator& operator=(CacheSimulator&&) = delete;
 
   virtual void Access(const BlockCacheTraceRecord& access);
-  void reset_counter() { sim_cache_->reset_counter(); }
-  double miss_ratio();
-  uint64_t total_accesses();
+  void reset_counter() {
+    num_misses_ = 0;
+    num_accesses_ = 0;
+    user_accesses_ = 0;
+    user_misses_ = 0;
+  }
+  double miss_ratio() {
+    if (num_accesses_ == 0) {
+      return -1;
+    }
+    return static_cast<double>(num_misses_ * 100.0 / num_accesses_);
+  }
+  uint64_t total_accesses() { return num_accesses_; }
+
+  double user_miss_ratio() {
+    if (user_accesses_ == 0) {
+      return -1;
+    }
+    return static_cast<double>(user_misses_ * 100.0 / user_accesses_);
+  }
+  uint64_t user_accesses() { return user_accesses_; }
 
  protected:
-  std::shared_ptr<SimCache> sim_cache_;
+  void UpdateMetrics(bool is_user_access, bool is_cache_miss);
+
+  std::unique_ptr<GhostCache> ghost_cache_;
+  std::shared_ptr<Cache> sim_cache_;
+  uint64_t num_accesses_ = 0;
+  uint64_t num_misses_ = 0;
+  uint64_t user_accesses_ = 0;
+  uint64_t user_misses_ = 0;
 };
 
 // A prioritized cache simulator that runs against a block cache trace.
 // It inserts missing index/filter/uncompression-dictionary blocks with high
 // priority in the cache.
 class PrioritizedCacheSimulator : public CacheSimulator {
  public:
-  PrioritizedCacheSimulator(std::shared_ptr<SimCache> sim_cache)
-      : CacheSimulator(sim_cache) {}
+  PrioritizedCacheSimulator(std::unique_ptr<GhostCache>&& ghost_cache,
+                            std::shared_ptr<Cache> sim_cache)
+      : CacheSimulator(std::move(ghost_cache), sim_cache) {}
+  void Access(const BlockCacheTraceRecord& access) override;
+
+ protected:
+  // Access the key-value pair and returns true upon a cache miss.
+  void AccessKVPair(const Slice& key, uint64_t value_size,
+                    Cache::Priority priority, bool no_insert,
+                    bool is_user_access, bool* is_cache_miss, bool* admitted,
+                    bool update_metrics);
+
+  Cache::Priority ComputeBlockPriority(const BlockCacheTraceRecord& access);
```

Comment:
Make it const?

Reply:
I added const for all functions that do not update member variables.